### PR TITLE
[FSDP] Clear deferred all-gather state after backward

### PR DESCRIPTION
**Pull request by sanketpurandare**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #196706
* #197488
* #196666
* #196665
* #196464
* #196463
* #196462
* #196956
* #196461
* __->__ #196460

## Problem

Implicit forward prefetch overlaps the next all-gather copy-in with the current
all-gather. The input is a view into the all-gather output, so FSDP retains the
current `AllGatherResult` until its copy-out has completed and the streams that
may reuse those buffers have been ordered after the copy-out event.

A partial grouped forward can miss the normal root post-forward flush. In the
chunked-head pattern, the group forward skips the head and the head later runs
standalone, leaving the final deferred result alive through backward. This PR
originally fixed that lifetime leak by releasing the shared state from root
post-backward.

The first implementation reused the forward-prefetch release operation at root
post-backward. That operation makes both high-priority all-gather streams wait
on the saved forward copy-out event before dropping the retained result. This
is the correct dependency when a subsequent forward all-gather is about to
reuse the buffers, but root post-backward has different ownership: it runs on
the current compute stream after the backward consumers have been enqueued.
Adding communication-stream waits at that boundary introduced a ROCm stream
dependency that prevented queued work from draining.

The pre-existing two-rank `test_multi_forward_module` exercises the same
FSDP-wrapped inner module twice per forward. On the ROCm MI350 distributed
shard, both ranks reached the final loss comparison and then timed out while
synchronizing outstanding GPU work. The same shard and test pass at the merge
base with the same CI image, and the failure reproduced on all three retries.

## Design

Make the cleanup boundary explicit with two internal operations:

| Boundary | Stream ordering before release |
| --- | --- |
| Forward replacement or root post-forward | The copy-in and all-gather streams wait on the saved copy-out event. |
| Root post-backward or iteration reset | The current compute stream waits on the saved copy-out event. |

`release_all_gather_state_for_comm_reuse()` owns the first policy. It is used
when the next implicit prefetch may immediately reuse the retained buffers on
either all-gather stream.

`release_all_gather_state_on_current_stream()` owns the second policy. It is
used after backward and during reset, where the current stream is the lifecycle
boundary. The next root forward already makes both all-gather streams wait on
the current stream before launching implicit prefetch, so this preserves the
required happens-before relationship without adding reverse dependencies at
post-backward.

Both operations enqueue GPU event waits only. There is no host synchronization,
backend-specific conditional, or caller-selected arbitrary wait set. Calling
either operation before lazy initialization remains safe and simply drops an
event-free retained state.

## Lifecycle Contract

| Caller | Cleanup operation | Reason |
| --- | --- | --- |
| `wait_for_unshard()` | Communication-stream reuse | A new forward all-gather may reuse the shared buffers immediately. |
| Root `_post_forward()` | Communication-stream reuse | Flush the final forward result while preserving safe reuse by the communication streams. |
| Root post-backward callback | Current-stream release | Backward owns the current stream; the next root forward reestablishes communication-stream ordering. |
| `reset_iter_state()` | Current-stream release | Exception recovery discards iteration state from the current-stream boundary. |

The root post-backward cleanup remains independent of `is_last_backward`.
Gradient accumulation should not retain a stale forward all-gather buffer
between backward calls, but it must release that state using the correct stream
boundary.

## Test Plan

- `python test/distributed/_composable/fsdp/test_fully_shard_comm.py TestFSDPCommContext`
- `CUDA_VISIBLE_DEVICES=0,1 python test/distributed/_composable/fsdp/test_fully_shard_training.py TestFullyShard1DTrainingCompose.test_partial_group_releases_deferred_all_gather_after_backward TestFullyShard1DTrainingCompose.test_reset_iter_state_after_exception TestFullyShard1DTrainingCore.test_multi_forward_module`
- `CUDA_VISIBLE_DEVICES=0,1 python test/distributed/_composable/fsdp/test_fully_shard_training.py TestFullyShard1DTrainingCore`
- Exact ROCm MI350 distributed shard containing `TestFullyShard1DTrainingCore.test_multi_forward_module`.
- `lintrunner --config=.lintrunner.toml torch/distributed/fsdp/_fully_shard/_fsdp_param_group.py torch/distributed/fsdp/_fully_shard/_fsdp_state.py test/distributed/_composable/fsdp/test_fully_shard_comm.py test/distributed/_composable/fsdp/test_fully_shard_training.py`

AI assistance disclosure:
> This change was authored with assistance from an AI coding tool and reviewed by the author.

### CODE REVIEW

**Comment by anijain2305** [comment]
Looks pretty good.

Two comments from AI assisted code review

* The test doesn't exercise the bug it fixes — test_fully_shard_training.py:2693. It fabricates AllGatherState(cast(Any, None), event) after forward already flushed the real state, so it only proves the new statement runs; it would pass against a bare all_gather_state = None with no stream waits, and never asserts the streams were ordered on the event. The      real leak paths are untested: a deferred state re-armed during backward AC recompute (_fsdp_param_group.py:487-497 sets it while _post_forward's flush is gated on iter_forward_root is self), and the chunked-loss/incomplete-root path behind _force_complete_incomplete_states.
* Gap: forward-only path still leaks. With an incomplete forward root under no_grad/inference, _post_forward's flush is skipped and no backward callback fires, so the buffer is retained until the next forward's copy-out. Same class of bug, uncovered.

**Comment by weifengpy** [comment]
is this PR really about AC? Or is it actually bufixing a partial forward bug ? meaning doing forward on model.norm, but not model.head, and ended up with dirty state

```
fully_shard([model.norm, model.head])
h = model(tokens, skip_head=True)  # Runs norm, skips head
for chunk in h.chunk(4, dim=1):
    logits = model.head(chunk)    # Runs head alone, skips norm
```

**Review by weifengpy — COMMENTED** [review]
see above

**Comment by sanketpurandare** [comment]
@claude provide a critical pr review, check for bugs, soundness of design, test coverage, redundant tests, slop code, cpmpactness of code, api design and provide feedback or suggest changes on how to make it better.

**Comment by sanketpurandare** [comment]
Replaced the synthetic `AllGatherState` fixture with the real grouped partial-forward path: a grouped norm/head forward skips the head, standalone chunked head calls genuinely leave deferred all-gather state armed, and root post-backward releases it. A separate helper-contract test proves that every requested stream waits on the recorded event before ownership is cleared. The implementation now centralizes release in `FSDPCommContext`, and the description explicitly scopes the fix to training iterations that reach root post-backward; forward-only partial execution may retain one bounded result until the next forward. The focused two-rank regression passes.

**Comment by sanketpurandare** [comment]
This is a partial grouped-forward lifecycle bug, not an AC-specific fix. I updated the title, description, and regression around the `fully_shard([norm, head])` pattern: the grouped forward skips the head, standalone head calls arm the deferred state, and the root post-backward boundary releases it. AC recomputation can reach the same stale-state boundary, but the PR no longer presents AC as the defining cause.

**Comment by sanketpurandare** [comment]
@claude Reworked the PR around the concrete grouped partial-forward defect. The test now genuinely arms deferred state through standalone head calls, proves it is live before backward, and proves root post-backward clears it; stream ordering is covered independently with recording event/stream doubles. The three release sequences now share `FSDPCommContext.release_all_gather_state()` while retaining each boundary's required wait streams. The fabricated state, inert `reshard_after_forward=False`, CUDA-graph test placement, hard-coded CUDA idiom, and AC-specific overclaims were removed. The PR explicitly retains the bounded forward-only limitation. The focused two-rank regression and helper test pass.

File: torch/distributed/fsdp/_fully_shard/_fsdp_state.py [hunk 1]
```diff
@@ -489,10 +494,7 @@ def _reset_iter_state(self) -> None:
                 "reset_iter_state must be called on the root FSDP module"
             )
         current_stream = self._device_handle.current_stream()
-        if ag_state := self._comm_ctx.all_gather_state:
-            if ag_state.event is not None:
-                current_stream.wait_event(ag_state.event)
-            self._comm_ctx.all_gather_state = None
+        self._comm_ctx.release_all_gather_state(current_stream)
```
Comment (fegin) [inline torch/distributed/fsdp/_fully_shard/_fsdp_state.py]: If we let `release_all_gather_sate` always wait for:
`self._comm_ctx.all_gather_copy_in_stream` and `self._comm_ctx.all_gather_stream`, do we still need sending current_stream? 

I think this wait it is more clear and self-contained.

**Review by fegin — APPROVED** [review]
Left one comment, it is not an accuracy blocker though, so stamped the PR.

**Comment by sanketpurandare** [comment]
@claude Moved the wait-set policy fully into `FSDPCommContext`: `release_all_gather_state()` is now argument-free and always orders both the copy-in and all-gather streams behind the copy-out event before releasing ownership. All release sites use that single contract, including `reset_iter_state()`, and the root-backward comment now explains why cleanup is independent of `is_last_backward`. The helper test asserts the exact two streams, while the real grouped partial-forward regression and `reset_iter_state()` call-path test prove the lifecycle boundaries. All three focused tests pass.

File: torch/distributed/fsdp/_fully_shard/_fsdp_state.py (hunk 1 above)
Comment (sanketpurandare) [reply torch/distributed/fsdp/_fully_shard/_fsdp_state.py]: Removed the `current_stream` argument. `release_all_gather_state()` now owns the complete wait set and always orders `all_gather_copy_in_stream` and `all_gather_stream`; callers no longer choose streams.

**Comment by sanketpurandare** [comment]
I used Codex to prepare this verification summary and reviewed it against the implementation and test results; it addresses the findings in this review: @claude Replaced the `reset_iter_state()` mock with outcome assertions that prove deferred state is armed and then released, consolidated pre-lazy-init handling into one stream guard, initialized `all_gather_state` at class scope, broadened the event docstring, enabled CPU coverage, and documented the partial-forward leak window. The focused communication-context and two distributed lifecycle tests pass, and changed-file lint is clean.
